Fusion frequency tables

kgaonkar6 opened the pull request:
🚨 merge after #50 
<!--Hi there, thanks for your contribution! Please take a moment to fill out this template to facilitate the review of your pull request.-->

### Purpose/implementation Section

#### What scientific question is your analysis addressing?

Using`FusionName`_`Fusion_Type` as a alteration ID we will generate tables with counts and frequencies per cancer_group and cohort. 

#### What was your approach?
The code is adapted from `https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/f70645b6c7e4eb15ea29e45e9ebf0adeb5798b9b/analyses/snv-frequencies` by @logstar 

Given a alteration dataframe with `Kids_First_Biospecimen_ID` and `Alt_ID` get_cg_ch_mut_freq_tbl() gets the counts and frequencies per cancer_group within cohort.


#### What GitHub issue does your pull request address?
https://github.com/PediatricOpenTargets/ticket-tracker/issues/70
https://github.com/PediatricOpenTargets/ticket-tracker/issues/72


### Directions for reviewers. Tell potential reviewers what kind of feedback you are soliciting.

#### Which areas should receive a particularly close look?
- `Reciprocal_exists` is for all fusions should this be restricted to Kinase?
-  `annots` and `Breakpoint Location` are not in putation-onco-fusion file should we add these back??

#### Is there anything that you want to discuss further?
Wanted to make sure, that the Gene_Symbol is separated as each gene fused in new row, so should there be columns specifying the position of the gene? Because it seems if we add Gene1A,Gene1B, Gene2A, Gene2B we have the same information in wide and also in long formation having a row per gene fused. 


#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
Yes


### Results

#### What types of results are included (e.g., table, figure)?
tables


#### What is your summary of the results?
The fusion calls ( currently it's PBTA only from `dev`) which we will updated once #42  goes in. 

#### Reproducibility Checklist

<!-- Check all those that apply or remove this section if it is not applicable.-->

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

<!-- Please review https://github.com/AlexsLemonade/OpenPBTA-analysis#documenting-your-analysis -->

- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

<!-- IF YOUR PULL REQUEST IS A DATA RELEASE, PLEASE REMOVE THE [HTML COMMENT TAG](https://html.com/tags/comment-tag/) FROM THE SECTION BELOW AND COMPLETE THE CHECKLIST-->

<!-- 
### Data Release Checklist
- [ ] Is the table in doc/data-file-descriptions.md up to date?
- [ ] Is doc/data-format.md up to date?
- [ ] Is doc/release-notes.md up to date?
- [ ] Is download-data.sh up to date?
- [ ] Was download-data.sh tested and did it complete without error?
-->

jharenza commented:
>Reciprocal_exists is for all fusions should this be restricted to Kinase?

yes

>annots and Breakpoint Location are not in putation-onco-fusion file should we add these back??

either yes, or create a put-onco-fusion file with annotations - probably the former to keep redundancy at a minimum

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,296 @@
+results_dir <- file.path(module_dir, "results")
+
+# Source function
+source(file.path(module_dir,"utils","freq_counts.R"))
```
Maybe move this to top? So readers could expect functions defined in other files at the beginning.

kgaonkar6 replied:
Thanks for pointing that out , I did check for some known genes for example BRAF,MET etc get and `ENSG00000157764` but that must be because of the `filter(!is.na(rmtl), !is.na(version)) ` so only a specific set of genes get annotated by ` ann_ensg_hugo_rmtl_df ` . I'll update that!

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,296 @@
+  stopifnot(identical(nrow(x), length(unique(x$Kids_First_Biospecimen_ID))))
+  
+  fx <- filter(x, !is.na(cancer_group), !is.na(cohort))
+  fx <- mutate(
+    fx, cancer_group_cohort = paste(cancer_group, cohort, sep = '___'))
```
```suggestion
```

The `cancer_group_cohort` column is no longer used in the latest commit of <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/45> .

Maybe delete them, and see if the results are still identical?

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,296 @@
+
+
+rm(tumor_kfbids)
+suppressMessages(gc(reset = TRUE))
```
```suggestion
```

`gc()` is no longer used in the latest commit of <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/45> .

I used it before to release RAM usage on my local computer.

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,296 @@
+      # if Alt_ID has no matched Patients_in_dataset
+      filter(!is.na(Patients_in_dataset))
+    
+    if ( nrow(res_df) != 0 ){
+    # merge fusion dataframe with the counts and frequencies in each cancer_group
+    # - per cohort 
+    # - per primary samples in cohort
+    # - per relapse samples in cohort
+    fusion_df <- fusion_df %>% 
+      select(-Kids_First_Biospecimen_ID,-Kids_First_Participant_ID) %>%
+      unique() %>%
+      inner_join(res_df,by="Alt_ID") 
+    
+    return(fusion_df)
+    }
```
```suggestion
    if ( nrow(res_df) != 0 ){
      # merge fusion dataframe with the counts and frequencies in each cancer_group
      # - per cohort 
      # - per primary samples in cohort
      # - per relapse samples in cohort
      fusion_df <- fusion_df %>% 
        select(-Kids_First_Biospecimen_ID,-Kids_First_Participant_ID) %>%
        unique() %>%
        inner_join(res_df,by="Alt_ID") 

      return(fusion_df)
    }
```

When `nrow(res_df) == 0`, it seems that `NULL` will be returned, will the following code still be able to run correctly?

kgaonkar6 replied:
since it's NULL it doesn't have anything to bind so it seems to work I can check again.

logstar commented on analyses/fusion-frequencies/README.md:
```diff
@@ -0,0 +1,65 @@
+
+Format the SNV mutation frequency table according to the latest spreadsheet that is attached in <https://github.com/PediatricOpenTargets/ticket-tracker/issues/64>.
+
+Merge the SNV mutation frequency tables of all `cancer_group_cohort`s.
```
```suggestion
Merge the fusion frequency tables of all `cancer_group_cohort`s.
```

logstar commented on analyses/fusion-frequencies/README.md:
```diff
@@ -0,0 +1,65 @@
+
+### Results
+
+Results are generated using PediatricOpenTargets/OpenPedCan-analysis data release v5.
```
```suggestion
Results are generated using PediatricOpenTargets/OpenPedCan-analysis data release v6.
```

logstar commented on analyses/fusion-frequencies/README.md:
```diff
@@ -0,0 +1,65 @@
+
+Results are generated using PediatricOpenTargets/OpenPedCan-analysis data release v5.
+
+`results/snv-consensus-annotated-mut-freq.tsv` is the merged Fusion frequency table for all `cancer_group_cohort`s
```
```suggestion
The merged fusion frequency table of all `cancer_group_cohort`s is output in TSV and JSON formats using `readr::write_tsv()` and `jsonlite::write_json()` respectively.

- `results/putative-oncogene-fusion-annotated-freq.json`
- `results/putative-oncogene-fusion-annotated-freq.tsv`
```

logstar commented on analyses/fusion-frequencies/run-frequencies.sh:
```diff
@@ -0,0 +1,23 @@
+ print dirname(abs_path(@ARGV[0]));' -- "$0")"
+cd "$script_directory" || exit
+
+# If results directory exists, remove the existing results.
+# If existing .gz files exist, gzip command asks for confirmation.
+# Design-wise, it is not surprising to remove previously generated results
+# before a new run.
+if [[ -d results ]]; then
+    rm -r results
+fi
```
```suggestion
```

Removed as suggested by @jharenza at <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/27#discussion_r664938856>.

logstar commented on analyses/fusion-frequencies/utils/freq_counts.R:
```diff
@@ -0,0 +1,226 @@
+      Frequency_in_overall_dataset = '',
+      Total_primary_tumors_mutated_Over_Primary_tumors_in_dataset = '',
+      Frequency_in_primary_tumors = '',
+      Total_relapse_tum60ors_mutated_Over_Relapse_tumors_in_dataset = '',
```
```suggestion
      Total_relapse_tumors_mutated_Over_Relapse_tumors_in_dataset = '',
```

logstar reviewed (changes requested):
Thank you for preparing this module @kgaonkar6 !

The code and documentation look good to me.

I wonder if you could add missing Ensembl (ENSG) IDs to the result tables by joining `ensg-hugo-rmtl-v1-mapping.tsv`, perhaps by revising the following procedure for adding RMTL. For example, the first line of the TSV file has gene symbol `LINC01019` and ENSG ID missing, but `ENSG00000248118   LINC01019` mapping exits in `ensg-hugo-rmtl-v1-mapping.tsv`.

<https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/4f4e7f0e0d006c077af6b716ac2a16a0efdd4138/analyses/fusion-frequencies/01-fusion-frequencies.R#L228-L241>

Adding ENSG IDs will also add more `Gene_full_name`s and `Protein_RefSeq_ID`, because they are joined by ENSG IDs.

<https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/4f4e7f0e0d006c077af6b716ac2a16a0efdd4138/analyses/fusion-frequencies/01-fusion-frequencies.R#L265-L268>

kgaonkar6 commented:
Thanks for the reviews !
 
@jharenza,I've update the reciprocal_exitsts to be TRUE if either 1 gene in (Gene1A or Gene1B) is kinase and will thus have "Yes" or "No" in `DomainRetainedGene1A` or `DomainRetainedGene1B`. 
I'll add the code to retain `annots` and add `Breakpoint location` code to the [script](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/dev/analyses/fusion_filtering/05-QC_putative_onco_fusion_distribution.Rmd) generating the putative-onco-fusion in 
 a staggered PR from #40 

 @logstar, thanks for the updates on comments and the missing ENSG ids, I've updated the gene id match code now.

kgaonkar6 commented:
@jharenza  the annots column is slightly difference in arriba annotation (uniquely has duplication/translocation/deletion values) and StarFusion annotation if Fusion is called in both. This will add another row uniqued by the values in `annots`, is that acceptable?

Example:
Arriba `["INTERCHROMOSOMAL[chr7--chr2]"],translocation`
StarFuson `[INTERCHROMOSOMAL[chr7--chr2]]`

Or should I look into aggregating it by creating a vector and unique it

jharenza commented:
> @jharenza the annots column is slightly difference in arriba annotation (uniquely has duplication/translocation/deletion values) and StarFusion annotation if Fusion is called in both. This will add another row uniqued by the values in `annots`, is that acceptable?
> 
> Example:
> Arriba `["INTERCHROMOSOMAL[chr7--chr2]"],translocation`
> StarFuson `[INTERCHROMOSOMAL[chr7--chr2]]`
> 
> Or should I look into aggregating it by creating a vector and unique it

I think aggregating all of those per fusion breakpoints/type would be better.

kgaonkar6 commented:
Just wanted add that the PR is updated with the aggregated `annots` aggreagted at non-caller specific columns in https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/50 as well as `Breakpoint location` is added.

kgaonkar6 commented:
Update to reduce the code being reviewed I am adding just the `FusionName+Fusion_Type` and `Fused Gene` frequencies in this PR as step1 .  

The additional annotation for ENSG ids, EFO, MONDO,etc was planned to be added to be another PR as step2 but as per slack convo we will have a common function to add the annotation by @logstar.

jharenza commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,242 @@
+                  "Reciprocal_exists_either_gene_kinase"="",
+                  "Gene1A_anno"="",
+                  "Gene1B_anno"="",
+                  "Gene2A_anno"="",
+                  "Gene2B_anno"="",
```
I am still seeing NAs in these two columns when I read this in:
```
fus <- read_tsv("~/Documents/GitHub/ped-ot/OpenPedCan-analysis/analyses/fusion-frequencies/results/putative-oncogene-fusion-freq.tsv", 
                na=character(0))
```
but the other columns are blanks, as expected. Could be because Gene2A_anno and Gene2B_anno are of class `logical` instead of `character`. Perhaps, you may need to apply @logstar's solution as in [this commit 38629de](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/45/commits/38629de5fa8fd6acb863056e9a46d8262558d790) for these columns.

kgaonkar6 replied:
Thanks for the review! 
I looked into the NAs in Gene2A_anno and Gene2B_anno, those should not be all NA. I also see it in `analyses/fusion_filtering/results/fusion-putative-oncogenic.tsv` so an upstream fix in fusion_filtering will be needed here. 

The code from @logstar is already part of the code ( basically all code here is adapted from the code from #45 )  unless I missed an update for the code.

jharenza replied:
I don't think we need to have this in the `fusion-putative-oncogenic.tsv` but only in the FNL-specific files - I prefer having NAs in files rather than blanks in general, but this is their specific request, so I think it should only apply to their files.

kgaonkar6 replied:
Right, I get that the NA to blank is specific for FNL files. 

I wanted to just mention that the original `fusion-putative-oncogenic.tsv` had Gene2A_anno and Gene2B_anno as all `NA` which is not expected  because oncogene like MET are Gene2A/B. So I went back and added `guess_max` to read the files correctly in #50 and now `fusion-putative-oncogenic.tsv` has non-NA values in `*_anno` columns. 

After I updated the #50 I also reran the code here and now seems to have spaces in `Gene2A_anno` and `Gene2B_anno` when I read it as `fus <- read_tsv("~/Documents/PedOpenTargets/OpenPBTA-analysis/analyses/fusion-frequencies/results/putative-oncogene-fusion-freq.tsv",na=character(0))`

```
> fus$Gene2A_anno %>% unique()
[1] ""                                                  "TranscriptionFactor"                              
[3] "Oncogene, Kinase"                                  "Oncogene"                                         
[5] "TumorSuppressorGene"                               "Kinase, TumorSuppressorGene, CosmicCensus, Kinase"
[7] "TumorSuppressorGene, Oncogene"                     "Kinase"                                           
[9] "Kinase, CosmicCensus, Oncogene, Kinase"         
```

Let me know if I missed anything.

jharenza replied:
OH, I see I see - gotcha. I misunderstood!

jharenza reviewed (changes requested):
Two minor change requests: two columns still contain NAs, so those need to be fixed.

For the `annots` column, can you clean up the annotations which start as `[], ` and just remove that piece for aesthetics in the OT tables?

Thanks!

jharenza reviewed (commented):
I hit submit before reviewing the fused-gene table!

I noticed that in the fused-gene table, you have duplicate gene rows, so that should be fixed. Though, did we talk about having this as a separate table? I did not have that one on my radar..

jharenza commented:
I also just spot-checked the frequencies for KIAA1549--BRAF, and something is not right here - we only have 1/285 LGG patients with this fusion? There is usually 60% of patients with this fusion, so can you check your logic for calculations?

jharenza commented:
> I also just spot-checked the frequencies for KIAA1549--BRAF, and something is not right here - we only have 1/285 LGG patients with this fusion? There is usually 60% of patients with this fusion, so can you check your logic for calculations?

Disregard this - I was looking at the reciprocal inframe fusion. This is all good!

kgaonkar6 commented:
> I hit submit before reviewing the fused-gene table!
> 
> I noticed that in the fused-gene table, you have duplicate gene rows, so that should be fixed. Though, did we talk about having this as a separate table? I did not have that one on my radar..

I think it was discussed in one of the meetings but I might be wrong. I can delete those files

jharenza commented:
> > I hit submit before reviewing the fused-gene table!
> > I noticed that in the fused-gene table, you have duplicate gene rows, so that should be fixed. Though, did we talk about having this as a separate table? I did not have that one on my radar..
> 
> I think it was discussed in one of the meetings but I might be wrong. I can delete those files

let's ask on Wednesday - I think the ask was to make sure we have each fusion designated by either gene, which you have in the fusion table.

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,242 @@
+
+print(alt_id)
+
+if (opt$alt_id == "FusionName,Fusion_Type") {
```
Might be more readable if tested on `alt_id`.

```suggestion
if (identical(alt_id, c("FusionName", "Fusion_Type"))) {
```

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,242 @@
+                 "Reciprocal_exists_either_gene_kinase",
+                 "Gene1A_anno","Gene1B_anno","Gene2A_anno","Gene2B_anno")
+  
+} else if (opt$alt_id == "Gene_Symbol") {
```
Might be more readable if tested on `alt_id`.

```suggestion
} else if (identical(alt_id, "Gene_Symbol")) {
```

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,242 @@
+
+if (opt$alt_id == "FusionName,Fusion_Type") {
+  fusion_df <- fusion_df %>%
+    mutate( Alt_ID=paste(!!as.name(alt_id),sep="_"))
```
I wonder if you could briefly explain how this works. I tried the following example in the docker image, but the results seem to be unexpected.
```text
> tmp <- tibble(a = c('a', 'b', 'c'), b = c('d', 'e', 'f'))
> tmp
# A tibble: 3 x 2
  a     b    
  <chr> <chr>
1 a     d    
2 b     e    
3 c     f    
> cols_to_paste <- c('a', 'b')
> tmp %>% mutate(paste_ab = paste(!!as.name(cols_to_paste), sep = '_'))
# A tibble: 3 x 3
  a     b     paste_ab
  <chr> <chr> <chr>   
1 a     d     a       
2 b     e     b       
3 c     f     c       
```

logstar replied:
Seems the results of the following expression is expected.

```text
> tmp %>%
+   mutate(paste_ab = paste(!!as.name(cols_to_paste[1]),
+                           !!as.name(cols_to_paste[2]),
+                           sep = '_'))
# A tibble: 3 x 3
  a     b     paste_ab
  <chr> <chr> <chr>   
1 a     d     a_d     
2 b     e     b_e     
3 c     f     c_f       
```

```suggestion
    mutate( Alt_ID=paste(!!as.name(alt_id[1]), !!as.name(alt_id[2]),sep="_"))
```

kgaonkar6 replied:
Thanks for pointing this out @logstar!

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,242 @@
+      fusion_df, td_htl_dfs$overall_htl_df, td_htl_dfs$primary_htl_df,
+      td_htl_dfs$relapse_htl_df, c_cancer_group, c_cohorts) %>%
+      # if Alt_ID has no matched Patients_in_dataset
+      filter(!is.na(Patients_in_dataset))
```
Could you briefly explain the scenario for `Patients_in_dataset` being NA? As `Patients_in_dataset` is determined by `length(unique(bpid_ss_htl_df$Kids_First_Participant_ID))`, it is odd that `length()` could return NA on a table column, even when `Alt_ID` has no matched `Kids_First_Participant_ID` in the independent sample list.

Relevant code links:

<https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/d7bc257f3fc501cadc65eaa2e24eede17e6591a8/analyses/fusion-frequencies/utils/freq_counts.R#L134-L142>

<https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/d7bc257f3fc501cadc65eaa2e24eede17e6591a8/analyses/fusion-frequencies/utils/freq_counts.R#L149>

<https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/d7bc257f3fc501cadc65eaa2e24eede17e6591a8/analyses/fusion-frequencies/utils/freq_counts.R#L184>

kgaonkar6 replied:
Hmm that's right `Patients_in_dataset` cannot be NA, checking

kgaonkar6 replied:
Removed the filter for NA for Patients_in_dataset

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,242 @@
+
+m_fus_freq_tbl <- m_fus_freq_tbl %>%
+  mutate(Dataset = if_else(Dataset == 'PBTA&GMKF&TARGET',
+                           true = 'all_cohorts', false = Dataset))
```
Currently, `all_cohorts` denotes not only `'PBTA&GMKF&TARGET'` but also any `Dataset` value that has more than one cohorts, e.g. `'PBTA&GMKF'` and `'GMKF&TARGET'`.

```suggestion
m_fus_freq_tbl <- m_fus_freq_tbl %>%
  mutate(Dataset = if_else(str_detect(Dataset, '&'),
                           true = 'all_cohorts', false = Dataset))
```

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,242 @@
+         Frequency_in_primary_tumors,
+         Total_relapse_tumors_mutated_Over_Relapse_tumors_in_dataset,
+         Frequency_in_relapse_tumors) %>%
+  unique() %>%
```
Could you explain the scenarios for having duplicates? I assume `Alt_ID` values are all unique, so `group_by(Alt_ID)` has all unique rows.

kgaonkar6 replied:
Ah yes there will not be duplicates here, we saw some duplicates at Fused_gene level files in a version which were genes fused in Neuroblastoma; this was resolved by adding `Dataset` so don't need to unique here.

logstar replied:
Thank you for looking into this issue.

I still found some duplicates in the results. The `uniq -c` counts unique lines, and the number of unique lines are listed at the beginning of each line. The `-d` option shows only duplicates.

```text
$ sort fusion-frequencies/results/putative-oncogene-fusion-freq.tsv  | uniq -c -d | head
      2 C11orf95--RELA  in-frame        C11orf95        Gene1A          Genic   [INTRACHROMOSOMAL[chr11:1.88Mb]], duplication                   FALSE           Oncogene, TranscriptionFactor                   CNS Embryonal tumor     PBTA    1/14    7.14%   0/9     0.00%   1/4     25.00%
      2 C11orf95--RELA  in-frame        C11orf95        Gene1A          Genic   [INTRACHROMOSOMAL[chr11:1.88Mb]], duplication                   FALSE           Oncogene, TranscriptionFactor                   Ependymoma      PBTA    25/86   29.07%  20/69   28.99%  7/19    36.84%
      2 C11orf95--RELA  in-frame        C11orf95        Gene1A          Genic   [INTRACHROMOSOMAL[chr11:1.88Mb]], duplication                   FALSE           Oncogene, TranscriptionFactor                   Glial-neuronal tumor NOS        PBTA    1/7     14.29%  0/5     0.00%   1/2     50.00%
      2 C11orf95--RELA  in-frame        RELA    Gene1B          Genic   [INTRACHROMOSOMAL[chr11:1.88Mb]], duplication                   FALSE           Oncogene, TranscriptionFactor  CNS Embryonal tumor      PBTA    1/14    7.14%   0/9     0.00%   1/4     25.00%
      2 C11orf95--RELA  in-frame        RELA    Gene1B          Genic   [INTRACHROMOSOMAL[chr11:1.88Mb]], duplication                   FALSE           Oncogene, TranscriptionFactor  Ependymoma       PBTA    25/86   29.07%  20/69   28.99%  7/19    36.84%
      2 C11orf95--RELA  in-frame        RELA    Gene1B          Genic   [INTRACHROMOSOMAL[chr11:1.88Mb]], duplication                   FALSE           Oncogene, TranscriptionFactor  Glial-neuronal tumor NOS PBTA    1/7     14.29%  0/5     0.00%   1/2     50.00%
      2 CCAR1--GRID1    frameshift      CCAR1   Gene1A          Genic   [INTRACHROMOSOMAL[chr10:16.81Mb]], inversion                    FALSE   TumorSuppressorGene                    Diffuse midline glioma   PBTA    1/55    1.82%   1/47    2.13%   0/11    0.00%
      2 CCAR1--GRID1    frameshift      GRID1   Gene1B          Genic   [INTRACHROMOSOMAL[chr10:16.81Mb]], inversion                    FALSE   TumorSuppressorGene                    Diffuse midline glioma   PBTA    1/55    1.82%   1/47    2.13%   0/11    0.00%
      2 LPA--QKI        frameshift      LPA     Gene1A          Genic   [INTRACHROMOSOMAL[chr6:2.75Mb]], inversion                      FALSE           TumorSuppressorGene, CosmicCensus                       CNS Embryonal tumor     PBTA    1/14    7.14%   1/9     11.11%  0/4     0.00%
      2 LPA--QKI        frameshift      QKI     Gene1B          Genic   [INTRACHROMOSOMAL[chr6:2.75Mb]], inversion                      FALSE           TumorSuppressorGene, CosmicCensus                       CNS Embryonal tumor     PBTA    1/14    7.14%   1/9     11.11%  0/4     0.00%
```

```text
$ sort fusion-frequencies/results/putative-oncogene-fused-gene-freq.tsv  | uniq -c -d | head
     10 7SK     Ependymoma      PBTA    5/86    5.81%   4/69    5.80%   1/19    5.26%
     10 7SK     High-grade glioma/astrocytoma   PBTA    1/84    1.19%   0/51    0.00%   0/33    0.00%
     10 7SK     Low-grade glioma/astrocytoma    PBTA    2/285   0.70%   1/237   0.42%   0/51    0.00%
     10 7SK     Medulloblastoma PBTA    1/118   0.85%   1/105   0.95%   0/13    0.00%
     10 7SK     Neuroblastoma   GMKF    4/344   1.16%   4/209   1.91%   0/0
     10 7SK     Neuroblastoma   all_cohorts     4/1042  0.38%   4/403   0.99%   0/10    0.00%
      2 AAK1    High-grade glioma/astrocytoma   PBTA    1/84    1.19%   1/51    1.96%   0/33    0.00%
      2 AATK    Neuroblastoma   GMKF    1/344   0.29%   1/209   0.48%   0/0
      2 AATK    Neuroblastoma   all_cohorts     1/1042  0.10%   1/403   0.25%   0/10    0.00%
```

The duplicates are probably introduced by the `inner_join`, as there are duplicated `Alt_ID`s in `fusion_df`. The mutation frequency table `res_df` has no duplicate, so the mutation frequencies were computed correctly. I was concerned that the `res_df` also has duplicates, which would imply the mutation frequencies are incorrect.

<https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/907dc82ca83a7172d213424ecb8335cb7a086125/analyses/fusion-frequencies/01-fusion-frequencies.R#L210-L213>

To remove the duplicates, the removed `unieuq() %>%` could be added back, or change the `inner_join` procedure to the following:

```R
      fusion_df <- fusion_df %>% 
        select(keep_cols, "Alt_ID") %>%
        unique() %>%
        inner_join(res_df,by="Alt_ID")
```

logstar commented on analyses/fusion-frequencies/README.md:
```diff
@@ -0,0 +1,68 @@
+- `../../data/histologies.tsv`
+- `../fusion_filtering/results/fusion-putative-oncogenic.tsv`
+- `../independent-samples/results/independent-specimens.wgs.primary.tsv`
+- `../independent-samples/results/independent-specimens.wgs.relapse.tsv`
```
Should we use `eachcohort` independent sample lists that are updated in the `independent-samples` module?

```suggestion
- `../independent-samples/results/independent-specimens.rnaseq.primary.tsv`
- `../independent-samples/results/independent-specimens.rnaseq.relapse.tsv`
```

kgaonkar6 replied:
ok updating!

logstar replied:
Thank you for updating!

logstar commented on analyses/fusion-frequencies/run-frequencies.sh:
```diff
@@ -0,0 +1,50 @@
+        --relapse_independence_list ../independent-samples/results/independent-specimens.rnaseq.relapse.tsv \
+        --output_filename "putative-oncogene-fused-gene-freq"
+        
+# gather additional annotation for Fusion level         
+Rscript 02-frequencies-annot.R --freq_file results/putative-oncogene-fusion-freq.tsv \
+        --efo_mondo_file ../../data/efo-mondo-map.tsv  \
+        --ensg_hugo_rmtl_file ../../data/ensg-hugo-rmtl-v1-mapping.tsv \
+        --output_filename "putative-oncogene-fusion-freq-annotated"
+
+# gather additional annotation for Fused Gene level           
+Rscript 02-frequencies-annot.R --freq_file results/putative-oncogene-fused-gene-freq.tsv \
+        --efo_mondo_file ../../data/efo-mondo-map.tsv  \
+        --ensg_hugo_rmtl_file ../../data/ensg-hugo-rmtl-v1-mapping.tsv \
+        --output_filename "putative-oncogene-fused-gene-freq-annotated"        
```
Maybe comment out these lines? As it is going to be revised, according to <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/49#issuecomment-880170225>.

logstar reviewed (changes requested):
Thank you for the updates @kgaonkar6 .

The module runs well in the docker image, and results look identical to the uploaded ones.

Following are my specific comments.

logstar commented:
I was also wondering if you could briefly explain Gene1A, Gene1B, Gene2A, and Gene 2B in the fusion-frequencies or fusion_filtering README.md. Also, what is the difference between (`ANTXR1--BRAF`, `[INTERCHROMOSOMAL[chr2--chr7]], translocation`) and (`BRAF--ANTXR1`, `[INTERCHROMOSOMAL[chr7--chr2]], translocation`)?

kgaonkar6 commented:
@logstar thanks for the detailed review! Looking into some of the un-resolved issues now and updating the code. 

About this:
> I was also wondering if you could briefly explain Gene1A, Gene1B, Gene2A, and Gene 2B in the fusion-frequencies or fusion_filtering README.md. 

Ok yes good point, will add a brief description here. 

> Also, what is the difference between (`ANTXR1--BRAF`, `[INTERCHROMOSOMAL[chr2--chr7]], translocation`) and (`BRAF--ANTXR1`, `[INTERCHROMOSOMAL[chr7--chr2]], translocation`)?

Arriba which calls such fusions considers each of the the fusion `ANTXR1--BRAF` and `BRAF--ANTXR1` as separate events that are reciprocal to each other. In one the ANTRXR1 is the 5` gene of the fused transcript and in another `BRAF` is the 5` gene of the fused transcript.

jharenza commented:
@kgaonkar6 - now that https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/long-format-table-utils is in, can you rerun this with v7 and add the RAPI annotations here or in a staggered PR? Then, convert to JSONL either here or in a staggered PR? Thanks!

kgaonkar6 commented:
Hi @jharenza @logstar I've updated the code to implement the annotation via annotator API and also added json files in my latest comments. 

Even though the RMTL is annotated by the  annotator api had to use `ensg-hugo-rmtl-mapping.tsv` to gather the ensg_id , let me know if there is a better way to do this. Thanks!

logstar commented:
> Hi @jharenza @logstar I've updated the code to implement the annotation via annotator API and also added json files in my latest comments.
> 
> Even though the RMTL is annotated by the annotator api had to use `ensg-hugo-rmtl-mapping.tsv` to gather the ensg_id , let me know if there is a better way to do this. Thanks!

@kgaonkar6 Thank you for the suggestion. I wonder if you have any suggestions on how to annotate RMTL without using the `ensg-hugo-rmtl-mapping.tsv`, so I could improve it accordingly.

kgaonkar6 commented:
> @kgaonkar6 Thank you for the suggestion. I wonder if you have any suggestions on how to annotate RMTL without using the `ensg-hugo-rmtl-mapping.tsv`, so I could improve it accordingly.

To clarify, I'm reading the `ensg-hugo-rmtl-mapping.tsv` to add the `ensg_id` column to the fusion calls. This might be an issue specific for fusion calls since we don't have ENSEMBL ids so wanted to mention it since it might be confusing why I'm reading `ensg-hugo-rmtl-mapping.tsv` in the code.

The RMTL annotation does in-fact come from your code directly so wasn't a comment on your code. Sorry for the confusion.

logstar commented:
> > @kgaonkar6 Thank you for the suggestion. I wonder if you have any suggestions on how to annotate RMTL without using the `ensg-hugo-rmtl-mapping.tsv`, so I could improve it accordingly.
> 
> To clarify, I'm reading the `ensg-hugo-rmtl-mapping.tsv` to add the `ensg_id` column to the fusion calls. This might be an issue specific for fusion calls since we don't have ENSEMBL ids so wanted to mention it since it might be confusing why I'm reading `ensg-hugo-rmtl-mapping.tsv` in the code.
> 
> The RMTL annotation does in-fact come from your code directly so wasn't a comment on your code. Sorry for the confusion.

Thank you for the clarification @kgaonkar6 ! Sorry that I misunderstood.

Regarding fusion calls that have no associated ENSG ID, I am also not sure how to better handle them. Thank you for pointing out this issue! I will keep this issue in mind when reviewing this PR.

For fusion calls that have associated ENSG IDs, I think adding the ENSG IDs with `ensg-hugo-rmtl-mapping.tsv` is the best practice now, because the module developers would be able to carefully handle module specific issues like the one you are encountering.

logstar commented on analyses/README.md:
```diff
@@ -23,6 +23,7 @@ Note that _nearly all_ modules use the harmonized clinical data file (`pbta-hist
 | [`focal-cn-file-preparation`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/focal-cn-file-preparation) | `pbta-cnv-cnvkit.seg.gz` <br> `pbta-cnv-controlfreec.tsv.gz` <br> `pbta-gene-expression-rsem-fpkm-collapsed.polya.rds` <br> `pbta-gene-expression-rsem-fpkm-collapsed.stranded.rds` <br> `analyses/copy_number_consensus_call/results/pbta-cnv-consensus.seg.gz` | Maps fr ...
 | [`fusion_filtering`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/fusion_filtering) | `pbta-fusion-arriba.tsv.gz` <br> `pbta-fusion-starfusion.tsv.gz` | Standardizes, filters, and prioritizes fusion calls | `results/pbta-fusion-putative-oncogenic.tsv`(included in data download) <br> `results/pbta-fusion-recurrent-fusion-byhistology.tsv` (included in data download) <br ...
 | [`fusion-summary`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/fusion-summary)| `pbta-histologies.tsv` <br> `pbta-fusion-putative-oncogenic.tsv` <br> `pbta-fusion-arriba.tsv.gz` <br> `pbta-fusion-starfusion.tsv.gz` | Generate summary tables from fusion files ([#398](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/398); [#623](https://github.com/AlexsLemonad ...
+| [`fusion-frequencies`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/fusion-frequencies) | `fusion_filtering/results/pbta-fusion-putative-oncogenic.tsv` <br> `independent-samples/results/independent-specimens.wgs.primary.tsv` <br> `independent-samples/results/independent-specimens.wgs.relapse.tsv` | Gather counts and frequencies for fusion per cancer_group and cohort | ...
```
```suggestion
| [`fusion-frequencies`](https://github.com/AlexsLemonade/OpenPBTA-analysis/tree/master/analyses/fusion-frequencies) | `fusion_filtering/results/fusion-putative-oncogenic.tsv` <br> `independent-samples/results/independent-specimens.rnaseq.primary.eachcohort.tsv` <br> `independent-samples/results/independent-specimens.rnaseq.relapse.eachcohort.tsv` | Gather counts and frequencies for fusion per cancer_group and cohort | None
```

`putative-oncogene-fusion-annotated-freq.json` and `putative-oncogene-fusion-annotated-freq.tsv` are not used by other modules.

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,287 @@
+         Frequency_in_primary_tumors,
+         Total_relapse_tumors_mutated_Over_Relapse_tumors_in_dataset,
+         Frequency_in_relapse_tumors) %>%
+  write_tsv(file.path(results_dir, paste0(output_filename,'.tsv')))
```
```suggestion
# write to tsv
annotated_m_fus_freq_tbl %>%
  select(keep_cols, Disease, MONDO, RMTL, EFO,Dataset,
         Total_alterations_Over_Patients_in_dataset,
         Frequency_in_overall_dataset,
         Total_primary_tumors_mutated_Over_Primary_tumors_in_dataset,
         Frequency_in_primary_tumors,
         Total_relapse_tumors_mutated_Over_Relapse_tumors_in_dataset,
         Frequency_in_relapse_tumors) %>%
  unique() %>%
  rename(Gene_symbol = Gene_Symbol) %>%
  write_tsv(file.path(results_dir, paste0(output_filename,'.tsv')))
```

Sorry the the confusion at <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/49#discussion_r672677078>. I meant to add this `unique()` back, as it was removed previously.

Currently, the results still have duplicated lines, because multiple columns that make rows unique are dropped in the `select` and `write_tsv` expression.

Also added `rename` to be consistent with the annotator. The annotator did not raise error, because gene symbols are not used to add RMTL or Gene_full_name, but other annotations may be added at a later point that requires gene symbols, so there would be no backward compatibility issue with the ETL.

logstar commented on analyses/fusion-frequencies/README.md:
```diff
@@ -0,0 +1,72 @@
+Each gene invloved in the fusion is annotated by a Gene_Position, for example:
+ - In a genic fusion where both breakpoints are within gene body the Gene_Position `Gene1A` will be the 5` gene and Gene1B will be 3` gene
+ - In an intergenic fusion where one or both breakpoint are outside the gene body, if the 5` breakpoint is a region between GeneX and GeneY  the Gene_Position of GeneX will be Gene1A, for GeneY will be Gene2A and their 3` breakpoint is within GeneE will be denoted as Gene_Position `Gene1B`.
+ - In an intergenic fusion where one or both breakpoint are outside the gene body, if the 3` breakpoint is within GeneE the Gene_Position of GeneE will be Gene1A and the 5` breakpoint is within GeneY and GeneZ, Gene Y will be denoted as Gene_Position `Gene1B` and GeneZ will be denoted as `Gene2B`.   
```
```suggestion
 - In a genic fusion where both breakpoints are within gene body the Gene_Position `Gene1A` will be the 5' gene and Gene1B will be 3' gene
 - In an intergenic fusion where one or both breakpoint are outside the gene body, if the 5' breakpoint is a region between GeneX and GeneY  the Gene_Position of GeneX will be Gene1A, for GeneY will be Gene2A and their 3' breakpoint is within GeneE will be denoted as Gene_Position `Gene1B`.
 - In an intergenic fusion where one or both breakpoint are outside the gene body, if the 3' breakpoint is within GeneE the Gene_Position of GeneE will be Gene1A and the 5' breakpoint is within GeneY and GeneZ, Gene Y will be denoted as Gene_Position `Gene1B` and GeneZ will be denoted as `Gene2B`.   
```

Thank you for the explanations @kgaonkar6 !

```text
Maybe change ` to ' to avoid GitHub from rendering monospace without the `.
```

logstar commented on analyses/fusion-frequencies/01-fusion-frequencies.R:
```diff
@@ -0,0 +1,287 @@
+                  "Gene2B_anno"="",
+                  "Fusion_anno"=""))%>%
+  # We want a single column that contains the gene symbols
+  tidyr::gather(Gene1A, Gene1B, Gene2A, Gene2B,
+                key = gene_position, value = gene_symbol) %>%
+  filter(!is.na(gene_symbol)) %>%
```
@jharenza @kgaonkar6 I looked into the issue that some gene symbols have no mapped ENSG ID as @kgaonkar6 pointed out at <https://github.com/PediatricOpenTargets/OpenPedCan-analysis/pull/49#issuecomment-889450851>. The following gene symbols are the only ones that have no ENSG ID: IGH-@, IGH@, IGL-@, and IGL@. (The code for checking them is listed in the end of this comment.) These gene symbols are [IGL immunoglobulin lambda locus](https://www.ncbi.nlm.nih.gov/gene?Db=gene&Cmd=DetailsSearch&Term=3535) and [IGH immunoglobulin heavy locus](https://www.ncbi.nlm.nih.gov/gene/3492), and there is no Ensembl link out from the NCBI gene pages. These gene symbols are introduced by starfusion, <https://github.com/STAR-Fusion/STAR-Fusion/issues/197#issuecomment-597215303>, but I am not sure if there is any difference between IGH-@ and IGH@, or between IGL-@ and IGL@.

These two loci are relevant to lymphoma, according to <https://www.sciencedirect.com/topics/biochemistry-genetics-and-molecular-biology/igl> and <https://www.sciencedirect.com/topics/biochemistry-genetics-and-molecular-biology/igh>, so I think we need to keep them searchable in the PedOT database.

I have been thinking about the following possible solutions:

- If starfusion also outputs actual IGH/IGL genes fusions, e.g. [IGHE ENSG00000211891](https://useast.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000211891;r=14:105597691-105601728) as additional rows, maybe we could remove these symbols. Also add note and code to check these are the only gene symbols without ENSG ID, for the ease of maintenance in the future. If starfusion does not output corresponding IGH/IGL genes, could we infer the genes from the break point or the other fused gene?
- If these IGH/L@ fusions really have no corresponding ENSG IDs and are useful, maybe we could discuss with the FNL team on handling these gene symbols that have no ENSG ID on the PedOT website in the `#ot-portal-content` channel.

```text
> f_tbl <- read_tsv('../../fusion-frequencies/results/putative-oncogene-fusion-freq.tsv.gz')
> f_gene_tbl <- read_tsv('../../fusion-frequencies/results/putative-oncogene-fused-gene-freq.tsv.gz')
> map_tbl <- read_tsv('../../../data/ensg-hugo-rmtl-mapping.tsv')
> f_tbl %>% filter(!Gene_Symbol %in% map_tbl$gene_symbol) %>% group_by(Gene_Symbol) %>% tally()
# A tibble: 4 x 2
  Gene_Symbol     n
  <chr>       <int>
1 IGH-@           3
2 IGH@           10
3 IGL-@           2
4 IGL@            2
> f_gene_tbl %>% filter(!Gene_Symbol %in% map_tbl$gene_symbol) %>% group_by(Gene_Symbol) %>% tally()
# A tibble: 4 x 2
  Gene_Symbol     n
  <chr>       <int>
1 IGH-@           9
2 IGH@           44
3 IGL-@           4
4 IGL@            4
```

kgaonkar6 replied:
Thanks @logstar from bringing up this issue.

We've noticed this issue with IGH@ symbols before and had to manually retain/annotate these fusions with IGH*@ as oncogenic fusions in OpenPBTA and current `fusion_filtering` modules. 

When we checked (last year I think) there were multiple IGH genes around the breakpoint and gtf, we don't have the IGH*@ genes in the gtf file we use to get the ENSG ID so you are right will have to discuss and manually add the matched ENSG ID to `ensg-hugo-rmtl-mapping.tsv`.

logstar replied:
> Thanks @logstar from bringing up this issue.
> 
> We've noticed this issue with IGH@ symbols before and had to manually retain/annotate these fusions with IGH*@ as oncogenic fusions in OpenPBTA and current `fusion_filtering` modules.
> 
> When we checked (last year I think) there were multiple IGH genes around the breakpoint and gtf, we don't have the IGH*@ genes in the gtf file we use to get the ENSG ID so you are right will have to discuss and manually add the matched ENSG ID to `ensg-hugo-rmtl-mapping.tsv`.

@kgaonkar6 Thank you for brining more context to this issue.

Regarding your suggested solution to manually add the matched ENSG IDs, I was wondering if there is any ENSG ID that directly maps to the IG*@ symbols. I checked NCBI gene, HGNC, mygene.info, and biomart, and there is no direct mapping from IGH/IGL to ENSG ID.

@jharenza and @kgaonkar6 Therefore, I was wondering if we should defer this issue as a fusion filtering enhancement and do not handle them in this `fusion-frequencies` module.

- IGH
  - NCBI gene: https://www.ncbi.nlm.nih.gov/gene/3492
  - mygene.info: http://mygene.info/v3/gene/3492
  - HGNC: https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:5477
  - biomart: see code below
- IGL
  - NCBI gene: https://www.ncbi.nlm.nih.gov/gene?Db=gene&Cmd=DetailsSearch&Term=3535
  - mygene.info: http://mygene.info/v3/gene/3535
  - HGNC: https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/HGNC:5853
  - biomart: see code below

```text
> bm_ensembl <- biomaRt::useEnsembl(
+   biomart = 'ensembl', dataset = 'hsapiens_gene_ensembl')
> biomaRt::getBM(
+   attributes = c('ensembl_gene_id', 'description'),
+   filters = c('hgnc_id'),
+   values = c(
+     'HGNC:5853', # IGL
+     'HGNC:5477', # IGH
+     'HGNC:5876' # this should have ENSG ID, to show this code works
+     ),
+   mart = bm_ensembl)
  ensembl_gene_id                                                            description
1 ENSG00000273737 immunoglobulin lambda variable 1-36 [Source:HGNC Symbol;Acc:HGNC:5876]
2 ENSG00000211655 immunoglobulin lambda variable 1-36 [Source:HGNC Symbol;Acc:HGNC:5876]
```

jharenza replied:
>@jharenza and @kgaonkar6 Therefore, I was wondering if we should defer this issue as a fusion filtering enhancement and do not handle them in this fusion-frequencies module.

Yes, we can do that to get this PR merged.

logstar replied:
> > @jharenza and @kgaonkar6 Therefore, I was wondering if we should defer this issue as a fusion filtering enhancement and do not handle them in this fusion-frequencies module.
> 
> Yes, we can do that to get this PR merged.

Thank you. I will submit a fusion-filtering ticket summarizing our discussions here and label the ticket with enhancement and low-priority.

I will resolve this conversation shortly.

logstar commented on analyses/fusion-frequencies/run-frequencies.sh:
```diff
@@ -0,0 +1,39 @@
+jq --compact-output '.[]' \
+  results/putative-oncogene-fusion-freq.json \
+  > results/putative-oncogene-fusion-freq.jsonl
+
```
```suggestion

rm results/putative-oncogene-fusion-freq.json
```

Maybe remove JSON, as they are not used.

logstar commented on analyses/fusion-frequencies/run-frequencies.sh:
```diff
@@ -0,0 +1,39 @@
+jq --compact-output '.[]' \
+  results/putative-oncogene-fused-gene-freq.json \
+  > results/putative-oncogene-fused-gene-freq.jsonl
+
```
```suggestion

rm results/putative-oncogene-fused-gene-freq.json
```

Maybe remove JSON, as they are not used.

logstar reviewed (changes requested):
Thank you for the updates @kgaonkar6 !

The updated module runs well, and the results are reproduced identically.

Following are my specific suggestions.

kgaonkar6 commented:
Thanks @logstar , I'll re-request review once I have rerun the module with your suggested changes.

jharenza commented:
>For fusion calls that have associated ENSG IDs, I think adding the ENSG IDs with ensg-hugo-rmtl-mapping.tsv is the best practice now, because the module developers would be able to carefully handle module specific issues like the one you are encountering.

Agree with this, thanks!

logstar reviewed (approved):
Thank you for the updates @kgaonkar6 !

The code updates since the last review look good to me.

The module runs well in the Docker image, and the results are reproduced identically. There is also no duplicated line in the result files anymore.

jharenza commented:
@kgaonkar6 I am still not seeing the ENSG id in the final tables - can you add this column to the tables?

kgaonkar6 commented:
@jharenza sorry about that, they are in the results files now.

jharenza reviewed (approved):
LGTM now!
